Add hosted runner and workflow file information retrieved from GitHub OIDC token

gr0 opened the pull request:
This PR aims to add support for hosted runner and workflow file information. The idea is to enhance the crafting scheme and prepare the `Runner` interface to be able to gather that information for each of the implementations that we have with a single implementation - GitHub for now, which is based on the SLSA GitHub generator and how it gets the data in a non falsifiable way. 

Any comments, suggestions or improvements ideas are more than welcome ;)

migmartri commented on pkg/attestation/crafter/runners/oidc/utils.go:
```diff
@@ -0,0 +1,103 @@
+//
+// Copyright 2023 The Chainloop Authors.
```
date 2025

migmartri commented on app/cli/internal/action/attestation_init.go:
```diff
@@ -229,6 +229,9 @@ func (action *AttestationInit) Run(ctx context.Context, opts *AttestationInitRun
 	}
 
+	// Resolve the hosted context information
+	action.c.ResolveHostedRunnerContext(ctx)
```
why do we need to run this now? What has changed?

jiparis replied:
`att init` vs `att push`. Which step do you think is best for running this?

gr0 replied:
I think the init state ain't bad, but looking for your input guys.

migmartri replied:
I think init it's fine, specially because eventually we should fail if the authentication is incorrect

gr0 replied:
Great, thanks.

migmartri commented on pkg/attestation/crafter/runner.go:
```diff
@@ -41,6 +42,12 @@ type SupportedRunner interface {
 	ID() schemaapi.CraftingSchema_Runner_RunnerType
+
+	// WorkflowFile returns the workflow file associated with this runner
+	WorkflowFile(ctx context.Context) string
```
we might need to be more explicit of what this should return, is this a reference? a path? the content?

gr0 replied:
Renamed, done.

migmartri commented on pkg/attestation/crafter/runners/githubaction.go:
```diff
@@ -16,16 +16,27 @@
-	return &GitHubAction{}
+	client, err := oidc.NewOIDCGitHubClient(context.Background())
+	if err != nil {
+		client = oidc.NewNoOPClient()
```
add a comment on why we need to do this. Maybe we should add an additional property in runner context called `runner_authenticated` that we can use to identify if the runner was authenticated using oidc or not (fallback)

migmartri commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,201 @@
+//
+// Copyright 2023 The Chainloop Authors.
```
```suggestion
// Copyright 2025 The Chainloop Authors.
```

gr0 replied:
Done.

migmartri commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,201 @@
+
+func (c *GitHubOIDCClient) decodeToken(token *oidc.IDToken) (*Token, error) {
+	var t Token
+	t.Issuer = token.Issuer
```
Do you need to do explicit assignment here? I believe this could be done by embedding the registered claims, otherwise you might be dropping some information other than issuer, audience and expiry?

https://github.com/chainloop-dev/chainloop/blob/6660e0bf45e3ece3619dd08929097125ddc774a4/internal/robotaccount/cas/robotaccount.go#L35

gr0 replied:
Refactored.

jiparis commented on app/controlplane/api/workflowcontract/v1/crafting_schema.proto:
```diff
@@ -40,6 +40,11 @@ message CraftingSchema {
   repeated PolicyGroupAttachment policy_groups = 7;
 
+  // Optional path to the workflow file content
+  string workflow_file = 8;
+  // Whether the runner is hosted or not
+  bool is_hosted_runner = 9;
```
I think these properties should be put in the CraftingState instead (crafting_state.proto), since it represents the current runtime data. CraftingSchema is meant to be used as a schema spec.

gr0 replied:
Thank you @jiparis - reworked. Appreciate that!

jiparis commented on pkg/attestation/crafter/crafter.go:
```diff
@@ -427,6 +427,13 @@ func (c *Crafter) ResolveEnvVars(ctx context.Context, attestationID string) erro
 
+func (c *Crafter) ResolveHostedRunnerContext(ctx context.Context) {
+	c.Logger.Debug().Bool("isHosted", c.Runner.IsHosted(ctx)).Msg("is runner hosted")
+	c.CraftingState.InputSchema.IsHostedRunner = c.Runner.IsHosted(ctx)
```
This is what I mean with the comment above. `CraftingState` is the object holding the current state. `InputSchema` is supposed to be immutable.

gr0 replied:
As mentioned in the other comment - done.

jiparis commented on pkg/attestation/crafter/runners/azurepipeline.go:
```diff
@@ -83,3 +84,11 @@ func (r *AzurePipeline) RunURI() (url string) {
+}
+
+func (r *AzurePipeline) IsHosted(_ context.Context) bool {
+	return false
```
Shouldn't it be `true` by default for all hosted CIs? (azure, github, gitlab, jenkins, circleci ...)

gr0 replied:
No, this shouldn't be `true` by default, in fact we should only set it to `true` if we are sure that this is hosted. Based on the current approach, so using environment variables ain't enough IMHO to say that something is hosted. Maybe we should rename this, for example to something like authenticated or something?

gr0 replied:
Actually, I went with the suggestion from @migmartri  and now we have `IsHosted` and `IsAuthenticated` - the `IsHosted` will be true for the defined runners, but the `IsAuthenticated` will be `true` only for GitHub Actions now where the OIDC token will be properly validated.

gr0 commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,201 @@
+
+	c = GitHubOIDCClient{
+		requestURL:  parsedURL,
+		bearerToken: os.Getenv(requestTokenEnvKey),
```
Good point, changed.

gr0 commented on pkg/attestation/crafter/runners/oidc/utils.go:
```diff
@@ -0,0 +1,103 @@
+)
+
+// Token represents the contents of a GitHub OIDC JWT token.
+type Token struct {
```
Good point, I'll rework it a bit.

gr0 commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,201 @@
+			return nil, err
+		}
+		return provider.Verifier(&oidc.Config{
+			SkipClientIDCheck: true,
```
Added a comment in the code.

gr0 commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,201 @@
+}
+
+// NewOIDCGitHubClient returns new GitHub OIDC provider client.
+func NewOIDCGitHubClient(_ context.Context) (Client, error) {
```
My bad during linter warning adjustments. Solved.

gr0 commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,201 @@
+}
+
+// Token requests an OIDC token from GitHub's provider, verifies it, and returns the token.
+func (c *GitHubOIDCClient) Token(ctx context.Context) (*Token, error) {
```
Done.

gr0 commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,201 @@
+}
+
+func (c *GitHubOIDCClient) requestToken(ctx context.Context, audience []string) ([]byte, error) {
+	req, err := http.NewRequest("GET", c.newRequestURL(audience), nil)
```
Done.

gr0 commented on pkg/attestation/crafter/runners/oidc/utils.go:
```diff
@@ -0,0 +1,103 @@
+	return &NoOPClient{}
+}
+
+func compareStringSlice(s1, s2 []string) bool {
```
Good point, done.

gr0 commented on pkg/attestation/crafter/runners/githubaction.go:
```diff
@@ -68,3 +79,19 @@ func (r *GitHubAction) RunURI() (url string) {
+}
+
+func (r *GitHubAction) IsHosted(ctx context.Context) bool {
+	token, err := r.oidcClient.Token(ctx)
```
Good point - done.

gr0 commented:
@migmartri @jiparis I refactored the code a bit, moved it around and added changes to attestation itself. When we agree on the approach I'll add unit tests to crafting.

jiparis commented on pkg/attestation/crafter/runners/githubaction.go:
```diff
@@ -68,3 +82,19 @@ func (r *GitHubAction) RunURI() (url string) {
+}
+
+func (r *GitHubAction) IsHosted(ctx context.Context) bool {
+	return r.oidcClient.IsHosted(ctx)
```
coming back to this, I think this should return `true` always, as the others. I cannot think of a case where this would be false for Github.

gr0 replied:
I agree to some extent, I can make it always `true`, but then exporting the right env variables in a local environment will make it also `true` right? Maybe I'm focused on that too much?

gr0 replied:
Changed :)

migmartri commented:
ptal at the test errors

migmartri commented on pkg/attestation/crafter/api/attestation/v1/crafting_state.proto:
```diff
@@ -150,6 +150,17 @@ message Attestation {
+  // Workflow file path that was used during build 
+  string workflow_file_path = 16;
+
+  string runner_environment = 17;
```
what does runner_environment mean?

gr0 replied:
This is what is returned in the OIDC token, for example `github-hosted`. This is what GitHub SLSA attestation action is also reporting.

migmartri replied:
can you add a comment then? Thanks!

gr0 replied:
Sure thing, done.

migmartri commented on pkg/attestation/crafter/crafter.go:
```diff
@@ -427,6 +431,15 @@ func (c *Crafter) ResolveEnvVars(ctx context.Context, attestationID string) erro
+func (c *Crafter) ResolveHostedRunnerContext(ctx context.Context) {
+	c.Logger.Debug().Bool("isHosted", c.Runner.IsHosted(ctx)).Msg("is runner hosted")
+	c.CraftingState.Attestation.IsHostedRunner = c.Runner.IsHosted(ctx)
+	c.Logger.Debug().Bool("isHosted", c.Runner.IsHosted(ctx)).Msg("is runner authenticated")
```
this log entry is strange, it seems to be talking both about is_hosted and authenticated

gr0 replied:
Yeah, copy+paste mistake, thanks for noticing. Fixed.

migmartri commented on app/cli/internal/action/attestation_init.go:
```diff
@@ -229,6 +229,9 @@ func (action *AttestationInit) Run(ctx context.Context, opts *AttestationInitRun
+	action.c.ResolveHostedRunnerContext(ctx)
+
 	// Load the env variables both the system populated and the user predefined ones
 	if err := action.c.ResolveEnvVars(ctx, attestationID); err != nil {
```
why instead of creating a new method `ResolveHostedRunnerContext` we make it part of `ResolveEnvVars`? renaming it of course i.e `ResolveRunnerInfo`

That way we only write in the state once

migmartri commented on pkg/attestation/crafter/runners/oidc/github_test.go:
```diff
@@ -0,0 +1,348 @@
+			}
+
+			// Test additional methods of the client interface
+			workflowPath := client.WorkflowFilePath(context.Background())
```
`context.Background()` is repeated multiple times, should we use a shared context or a test one?

gr0 replied:
Done.

migmartri commented on pkg/attestation/crafter/crafter.go:
```diff
@@ -418,6 +422,9 @@ func (c *Crafter) ResolveEnvVars(ctx context.Context, attestationID string) erro
 	}
 
+	// Resolve runner information
+	c.ResolveRunnerInfo(ctx)
```
I do not think you need to expose this method anymore right?

gr0 replied:
Good point, thanks, done.

migmartri commented on pkg/attestation/crafter/crafter.go:
```diff
@@ -370,6 +370,10 @@ func initialCraftingState(cwd string, opts *InitOpts) (*api.CraftingState, error
 			Head:                   headCommitP,
 			BlockOnPolicyViolation: opts.BlockOnPolicyViolation,
 			SigningOptions:         &api.Attestation_SigningOptions{TimestampAuthorityUrl: tsURL},
+			WorkflowFilePath:       opts.Runner.WorkflowFilePath(context.Background()),
```
we should probably get the context as parameter or at least store it in a variable

gr0 replied:
Ah I missed it, thank you!

migmartri commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,271 @@
+}
+
+// NewOIDCGitHubClient returns new GitHub OIDC provider client.
+func NewOIDCGitHubClient() (Client, error) {
```
this should return a concrete type not an interface https://duncanleung.com/go-idiom-accept-interfaces-return-types/

migmartri commented on pkg/attestation/crafter/runners/oidc/utils.go:
```diff
@@ -0,0 +1,74 @@
+
+// Client is the interface for an OIDC client.
+type Client interface {
+	WithAudience(audience []string) Client
```
it's strange that we have a mutator here while the rest seems to have properties. 

I'd move WithAudience to a optional opt in the constructor

gr0 replied:
Done.

migmartri commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,271 @@
+	return true
+}
+
+func (c *GitHubOIDCClient) RunnerEnvironment(ctx context.Context) string {
```
this error swalling is strange, shouldn't we return an error?

gr0 replied:
I was thinking about it when I was creating those functions in the `GitHubOIDCClient` - in general those errors are not actionable, at least at this point. We can of course return them, I did the change, but now the question is, should we propagate it further from `githubaction.go`, if yes - should we fail the attestation process if there is an error or should we just return empty values for workflow file and runner environment? WDYT?

gr0 replied:
Done.

gr0 commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,271 @@
+}
+
+// NewOIDCGitHubClient returns new GitHub OIDC provider client.
+func NewOIDCGitHubClient() (Client, error) {
```
Done.

migmartri commented on pkg/attestation/crafter/runners/daggerpipeline.go:
```diff
@@ -56,3 +57,19 @@ func (r *DaggerPipeline) RunURI() string {
+	return false
+}
+
+func (r *DaggerPipeline) RunnerEnvironment(_ context.Context) string {
```
env, UNKNOWN, SELF_HOSTED, MANAGED

gr0 replied:
Done.

migmartri commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,277 @@
+	RunnerEnvironment string `json:"runner_environment"`
+
+	// RawToken is the unparsed oidc token.
+	RawToken string
```
do you need to expose this token?

gr0 replied:
I removed that.

migmartri commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,277 @@
+
+// NewOIDCGitHubClient returns new GitHub OIDC provider client.
+func NewOIDCGitHubClient(opts ...Option) (*GitHubOIDCClient, error) {
+	var c GitHubOIDCClient
```
I'd initialize opts at the beginning

gr0 replied:
Done.

migmartri commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,277 @@
+
+	tokenBytes, err := c.requestToken(ctx, c.audience)
+	if err != nil {
+		return nil, err
```
you might want to start wrapping errors if you see fit

jiparis replied:
+1

gr0 replied:
Done.

migmartri commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,277 @@
+
+// WithAudience is deprecated. Use NewOIDCGitHubClient with WithAudience option instead.
+// This method is kept for backward compatibility.
+func (c *GitHubOIDCClient) WithAudience(audience []string) *GitHubOIDCClient {
```
this is not needed anymore?

migmartri commented on pkg/attestation/crafter/runners/oidc/github.go:
```diff
@@ -0,0 +1,277 @@
+	return &c, nil
+}
+
+func (c *GitHubOIDCClient) IsHosted(_ context.Context) bool {
```
this should not have the implementation here, 

Basically, this package is just a oidc client

migmartri commented on pkg/attestation/crafter/runners/githubaction.go:
```diff
@@ -68,3 +85,21 @@ func (r *GitHubAction) RunURI() (url string) {
 }
+
+func (r *GitHubAction) RunnerEnvironment(ctx context.Context) string {
+	runnerEnv, _ := r.oidcClient.RunnerEnvironment(ctx)
```
instead of calling the `RunnerEnvironment` just implement it here

gr0 commented on pkg/attestation/crafter/api/attestation/v1/crafting_state.proto:
```diff
@@ -150,6 +150,18 @@ message Attestation {
   SigningOptions signing_options = 15;
 
+  // Workflow file path that was used during build 
+  string workflow_file_path = 16;
```
Done.

migmartri commented on pkg/attestation/crafter/crafter.go:
```diff
@@ -370,6 +370,13 @@ func initialCraftingState(cwd string, opts *InitOpts) (*api.CraftingState, error
 			Head:                   headCommitP,
 			BlockOnPolicyViolation: opts.BlockOnPolicyViolation,
 			SigningOptions:         &api.Attestation_SigningOptions{TimestampAuthorityUrl: tsURL},
+			RunnerEnvironment: &api.RunnerEnvironment{
```
nice

migmartri commented on pkg/attestation/crafter/api/attestation/v1/crafting_state.proto:
```diff
@@ -150,12 +150,33 @@ message Attestation {
+  string environment = 2;
+
+  // Whether the runner is authenticated, i.e. via the OIDC token
+  bool is_authenticated_runner = 3;
```
```suggestion
  bool authenticated = 3;
```

migmartri commented on pkg/attestation/crafter/api/attestation/v1/crafting_state.proto:
```diff
@@ -150,12 +150,33 @@ message Attestation {
+  bool is_authenticated_runner = 3;
+
+  // Runner type
+  string type = 4;
```
Is this typed by any chance? Shoul'd this be? workflowcontract.v1.CraftingSchema.Runner.RunnerType

gr0 replied:
Good point, done.

migmartri commented on pkg/attestation/crafter/runner.go:
```diff
@@ -41,12 +42,21 @@ type SupportedRunner interface {
+	IsAuthenticated() bool
+
+	// RunnerEnvironment returns the runner environment
+	RunnerEnvironment() runners.RunnerEnvironment
```
```suggestion
	Environment() runners.RunnerEnvironment
```

since Runner is already the name of the interface

migmartri commented on pkg/attestation/crafter/runners/githubaction.go:
```diff
@@ -16,16 +16,48 @@
+	client, err := oidc.NewGitHubClient(oidc.WithAudience(defaultAudience))
+	if err != nil {
+		return &GitHubAction{
+			githubToken: nil,
```
do we have any logger that we could inject here so we can `debug` the fact that the github runner detection is unauthenticated for example?

gr0 replied:
Runners are created at the point where we don't have a logger anywhere, at least I don't see it.

migmartri commented on pkg/attestation/crafter/runners/oidc/utils.go:
```diff
@@ -0,0 +1,43 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oidc
```
I'd rename `utils.go` for `oidc` to be consistent with the package name

gr0 replied:
Good idea, done.

migmartri commented on pkg/attestation/renderer/chainloop/chainloop.go:
```diff
@@ -86,6 +86,10 @@ type ProvenancePredicateCommon struct {
+	// Additional properties related to runner
+	RunnerEnvironment   string `json:"runnerEnvironment,omitempty"`
+	RunnerAuthenticated bool   `json:"runnerAuthenticated,omitempty"`
+	WorkflowFilePath    string `json:"workflowFilePath,omitempty"`
```
```suggestion
	RunnerWorkflowFilePath    string `json:"RunnerWorkflowFilePath,omitempty"`
```

migmartri reviewed (commented):
Almost there, thanks!!

javirln commented on pkg/attestation/crafter/runner.go:
```diff
@@ -41,12 +42,21 @@ type SupportedRunner interface {
 
 var RunnersMap = map[schemaapi.CraftingSchema_Runner_RunnerType]SupportedRunner{
-	schemaapi.CraftingSchema_Runner_GITHUB_ACTION:   runners.NewGithubAction(),
+	schemaapi.CraftingSchema_Runner_GITHUB_ACTION:   runners.NewGithubAction(context.Background()),
```
I’m not sure how you see it, but to avoid waiting indefinitely, how about adding a context with a specific timeout—say, 15 seconds—so the request to the client doesn’t hang?

jiparis commented on pkg/attestation/crafter/runners/githubaction.go:
```diff
@@ -68,3 +100,28 @@ func (r *GitHubAction) RunURI() (url string) {
+	if r.githubToken != nil {
+		return r.githubToken.JobWorkflowRef
+	}
+	return ""
```
check https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#default-environment-variables for the default values for GITHUB_WORKFLOW_REF and RUNNER_ENVIRONMENT

gr0 replied:
as mentioned, it seem those env are not available for each run, at least I don't see them present

migmartri commented:
@gr0 ptal at the DCO error, you need to make sure all commits have signoff so please rebase them.

On another hand, what's left of this feature? I am thinking of cutting a new release and want to decide if we should wait for this or not.

gr0 commented:
@migmartri I'll do the proper sign off later today, I might have missed it. When it comes to what is left - it is tested and done :)

migmartri commented:
Ok, then feel free to merge it :)

Thanks
